site/content/blog/multi-env-deployment-ytt-kapp.md

```diff
@@ -0,0 +1,331 @@
+---
+title: "Multi environment deployment with ytt and kapp"
+slug: multi-env-deployment-ytt-kapp
+date: 2022-03-10
+author: Yash Sethiya
+excerpt: "Handle multiple environments with ytt and kapp"
+image: /img/logo.svg
+tags: ['carvel', 'kapp', 'ytt']
+
+---
+
+One of the most typical challenges when deploying a complex application is the handling of different deployment environments during the software lifecycle.
+
+The most typical setup is the trilogy of QA/Staging/Production environments. An application developer needs an easy way to deploy to the different environments and also to understand what version is deployed where.
+
+Unlike [many other tools used for templating](https://carvel.dev/ytt/docs/v0.40.0/ytt-vs-x/), ytt takes a different approach to work with YAML files. Instead of interpreting YAML configuration as plain text, it works with YAML structures such as maps, lists, YAML documents, scalars, etc. By doing so ytt is able to eliminate a lot of problems that plague other tools (character escaping, ambiguity, indentation, etc.). Additionally ytt provides Python-like language (Starlark) that executes in a hermetic environment making it friendly, yet more deterministic compared to just using general purpose languages directly or non-familiar custom templating languages.
+
+
+## How to handle different environment configurations
+
+This problem is typically solved in two ways: templating or patching. [ytt](https://carvel.dev/ytt) supports both approaches. In this section, we’ll see how ytt allows to template YAML configuration, and how it can patch YAML configuration via overlays.
+
+### Data values
+Data values provide a way to inject input data. In ytt, before a Data Value can be used in a template, it must be declared. This is typically done via [Data Values Schema](https://carvel.dev/ytt/docs/v0.40.0/how-to-write-schema/#overview). We will have a common schema file for all the environments.
+
+```yaml
+#! schema.yaml (#! is used for comment in ytt)
+#@data/values-schema
+---
+replicaCount: 1
+
+image:
+  name: nginx
+  tag: "1.14.2"
+
+nameOverride: ""
+fullnameOverride: ""
+
+appMode: ""
+certificatePath: ""
+databaseUser: ""
+databasePassword: ""
+```
+
+While normally schema defines the default configuration, it makes sense to create different values files for all different environments. Let's suppose we have two environments `staging` and `prod`.
```

Comment:
```suggestion
A schema sets the defaults for each data value as they are declared. We will want to override some of those defaults for each of our environments. We do this by creating a values file for each environment. Let's suppose we have two environments `staging` and `prod`.
```

Rationale: I'm picking up some sentence structure that lands for me as a kind of choice or trade-off: "maybe set a value in another file? maybe something else?". However, I believe your intent is to walk the user through _the_ process: first, declare the variables (data values); then tailor those values on a per environment basis. This suggestion is an attempt to sound more directive/definitive.